### ISSUE TITLE
IMAP or SMTP authentication

### ISSUE THREAD

**Issue by ltning** [rickbergfalk/sqlpad#633]
Having the ability to authenticate users against an IMAP or SMTP back-end would be very useful, especially in combination with the whitelist domain feature. If implemented, there should be a way to set default role (or even access) for users created automatically.

**Comment by rickbergfalk** [comment]
I'm not too familiar with using IMAP or SMTP for auth. 

Would this be like Slack's "magic link" where a user gets an email with a token to log in without a password?

**Comment by ltning** [comment]
No, it would simply mean talking to an SMTP or IMAP server and authenticating against it. Successful authentication would trigger creation of a barebone account with the default settings. It probably means you'd have to ask for an additional piece of information - the SMTP/IMAP username - if using that authentication back-end. It should only be allowed for users in the whitelisted domains, or better, domains assigned to an SMTP or IMAP auth back-end.

For us, either IMAP or SMTP is sufficient - implementing both options is probably overkill, but I can imagine there are scenarios where one would prefer one over the other. Either way, TLS (IMAP, SMTP) and STARTTLS (SMTP) must be supported, and specifying a list of trusted CAs (in a bundle file, as installed by the ca_root_nss package on most platforms) must be possible. Default should be to require validation of the certificate chain and hostname. It should be possible to disable chain validation and hostname validation independently.

Hope this helps :)

**Comment by rickbergfalk** [comment]
Ah ok -- it does. It doesn't look like the passportjs auth ecosystem has any built-in strategies for this, but a custom one could be built using this lib by chance for IMAP at least https://imapflow.com/

**Comment by eladeyal-intel** [comment]
Maybe LDAP would be easier to implement. I'm not sure that using SMPT or IMAP as authentication endpoints is best practice. http://www.passportjs.org/packages/passport-ldapauth/

**Comment by rickbergfalk** [comment]
I was wondering about that. @ltning is your SMTP/IMAP setup driven by LDAP or Active Directory by chance?

Using SMPT/IMAP auth isn't common, but is the concept itself dangerous? There's a part of me that thinks it is kind of a creative approach.

However... I'm guessing any sizable org is driving their email off of LDAP, so LDAP is used for the authentication mechanism instead of using the email server (which may in turn be forwarding the auth to the LDAP server).

**Comment by ltning** [comment]
@rickbergfalk No, we've been able to avoid that for 15 years, not going to start now :)

On a serious note; there is no inherent security issue in doing it "our way" though it's certainly not terribly common. We have, however, been using our internal IMAP server as the source of user accounts across our infrastructure for a long time. We modified the Gitlab integration in Mattermost to use IMAP back-end, Owncloud is using the IMAP server as back-end, we've done the same with a number of other internal support systems - SQLPad is missing (for now).

I don't know what you mean by "sizable", but we're >60 people across 10+ countries so it's no back yard operation. While using LDAP would certainly add some benefits and flexibility, it's also a lot harder to get "right", particularly security-wise, and the features we'd want from it would tend to either be commercial or have to be homegrown. We have a strong philosophy of "open source only" so all those enterprise-y features in some of those projects, which usually comes in the form of binary blobs for platforms we don't use, are a no-go.

**Comment by eladeyal-intel** [comment]
https://www.npmjs.com/package/passport-imap ?

Published 6 years ago

**Comment by rickbergfalk** [comment]
@ltning Sorry sizable was a poor choice of wording. I don't blame you for avoiding LDAP. I've also thought about adding support for it but have also avoided it 😅 It isn't something I'd want to tackle unless I personally was involved with an LDAP setup. 

@eladeyal-intel good find on the passport strategy. I searched the passport website but it wasn't listed. I'll have to take a look at the package.
